libsmtutil/Z3CHCInterface.cpp

```diff
@@ -103,8 +97,119 @@ pair<CheckResult, vector<string>> Z3CHCInterface::query(Expression const& _expr)
 	catch (z3::exception const&)
 	{
 		result = CheckResult::ERROR;
-		values.clear();
+		cex = {};
+	}
+
+	return {result, cex};
+}
+
+void Z3CHCInterface::setSpacerOptions(bool _preProcessing)
+{
+	// Spacer options.
+	// These needs to be set in the solver.
+	// https://github.com/Z3Prover/z3/blob/master/src/muz/base/fp_params.pyg
+	z3::params p(*m_context);
+	// These are useful for solving problems with arrays and loops.
+	// Use quantified lemma generalizer.
+	p.set("fp.spacer.q3.use_qgen", true);
+	p.set("fp.spacer.mbqi", false);
+	// Ground pobs by using values from a model.
+	p.set("fp.spacer.ground_pobs", false);
+
+	// Spacer optimization should be
+	// - enabled for better solving (default)
+	// - disable for counterexample generation
+	p.set("fp.xform.slice", _preProcessing);
+	p.set("fp.xform.inline_linear", _preProcessing);
+	p.set("fp.xform.inline_eager", _preProcessing);
+
+	m_solver.set(p);
+}
+
+/**
+Convert a ground refutation into a linear or nonlinear counterexample.
+The counterexample is given as an implication graph of the form
+`premises => conclusion` where `premises` are the predicates
+from the body of nonlinear clauses, representing the proof graph.
+
+This function is based on and similar to
+https://github.com/Z3Prover/z3/blob/master/src/muz/spacer/spacer_context.cpp#L2942
```

Comment:
@ekpyron added this comment, hopefully it makes it better. The mentioned Spacer function is similar to this one, modified here for nonlinear counterexamples.

Reply:
Maybe better to refer it by commit hash? ``master`` will change and this might point elsewhere in the future...

Reply:
Done